### ISSUE TITLE
Store samples in matrix rather than dataframe

### ISSUE THREAD

**Issue by aaronjg** [stan-dev/rstan#536]
Allocating very wide data.frames can take several hours. Though, data.frames take roughly the same amount of memory as matrices, they are much slower to allocate, create many more objects, and slow down garbage collection.

It would be nice to store the samples as matrices to make things faster. I am have a branch that has read_stan_csv create a matrix rather than a data.frame, while maintaining support for previous data.frame based stan.fit objects. 

For a 10,000 parameter model, this is an additional 15% speed improvement in reading in the data  over https://github.com/stan-dev/rstan/pull/532, and I believe would offer even greater speed improvements for larger models.

I'm not sure why things were initially created using data frames, but if using matrices makes sense, I can create a pull request.

https://github.com/aaronjg/rstan/tree/use_matrices

**Comment by bgoodri** [comment]
We are going to have to switch to numeric arrays eventually, but this requires changes to the C++ and would break anything that assumes the stanfit object is storing the draws as a list of list of scalars.

**Comment by aaronjg** [comment]
Are there changes in the C++ needed beyond https://github.com/aaronjg/rstan/commit/4c01df93a015a08965758ce459b68b68d56c08a9?  That fixes the summary function so that it calculate Neff and Rhat correctly.  I believe the stanfit object is storing the draws from a chain as a list of vectors currently, no?

I wasn't aware that the numeric arrays was on the roadmap already. Are there other planned features that depend on it?

**Comment by bgoodri** [comment]
List of vectors or data.frame we have the same issue, plus the occasional
mystery segfault on the C++ side where things are stored as nested
std::vectors.

**Comment by aaronjg** [comment]
Right, data.frames are lists of vectors with some extra attributes. You had mentioned that the stanfit object is storing the draws as a list of list of scalars, though. Are these being converted somewhere from lists of lists to lists of vectors? 

Also, what I have so far is really only a partial fix: storing data from csvs as matrix, and the augmenting the necessary functions to work with both the current data.frames/lists system as well as the matrix system. 

In that branch, the objects are still stored in the old system when the stanfit object is generated from sampling.

If there is interest, I could work on changing the C++ code further to use matrices throughout. It looks like this might only involve changing values.hpp?

In my ideal case, the draws could be stored in binary on disk as a memory mapped object, and then there would be no conversion from CSV needed, and unused parameters would never needed to be loaded into memory. But I'm not sure where that fits in the larger roadmap. It sounded at one point like there is some talk of getting more binary storage formats supported though.

**Comment by maverickg** [comment]
If matrix works well, I think it's a good idea.  Whatever data structure gets used, it needs to be persistent in R.  That is, if you save an stanfit object and load it next time, you still have the samples.

**Comment by aaronjg** [comment]
@maverickg I wasn't suggesting the mmap as the default. However, it would be very nice to have that as optional functionality so that the entire parameter set doesn't need to be loaded into memory at once. I would set it up so the default would be a matrix, and the user could optionally write out to a mmap object, similar to how they can write out to a CSV file now. If the user wanted to serialize the data and move it to another computer they would be responsible for copying the extra files, just like they would need to copy multiple CSV files now. 

If this sort of mmap interface wouldn't be used beyond me, I'd just leave it on my own fork.

Currently, I have it set up to use matrices when reading in from csv, with some changes in the C++ code to maintain compatibility with the current list of vectors generated from Rstan.
https://github.com/aaronjg/rstan/tree/use_matrices

@bgoodri Do you have any examples of models that cause segfaults from the nested std::vectors? I haven't run into this before and would like to use it as a test case.
